### PR TITLE
feature/lock

### PR DESCRIPTION
**Pull request by rgerrans**
I've tried to make all the changes on the back end to enable the creation of a Lock node. Please let me know if I got everything this time or missed any.

Thanks,

Ron

### CODE REVIEW

**Comment by rgerrans** [comment]
@andrei-tatar I've added a fan as well as the lock to the backend services. When you get a chance to update, I'll start testing both. Thanks

File: src/http/services/execute.service.ts [hunk 1]
```diff
@@ -47,10 +47,12 @@ export class ExecuteService {
         switch (execution.command) {
           case ExecuteCommandTypes.Brightness:
           case ExecuteCommandTypes.OnOff:
+		  case ExecuteCommandTypes.LockUnlock:
           case ExecuteCommandTypes.ColorAbsolute:
           case ExecuteCommandTypes.ThermostatTemperatureSetpoint:
           case ExecuteCommandTypes.ThermostatTemperatureSetRange:
           case ExecuteCommandTypes.ThermostatSetMode:
+		  case ExecuteCommandTypes.FanSpeed:
```
Comment (andrei-tatar) [inline src/http/services/execute.service.ts]: shouldn't this be `case ExecuteCommandTypes.SetFanSpeed`? Is it building?

File: src/http/services/query.service.ts [hunk 2]
```diff
@@ -65,6 +65,17 @@ export class QueryService {
             case 'garage':
                 state.openPercent = device.state.openPercent;
                 break;
+			case 'lock':
+				state.isLocked = device.state.isLocked;
+				state.isJammed = device.state.isJammed;
+				break;
+			case 'fan':
+				state.on - device.state.on;
```
Comment (andrei-tatar) [inline src/http/services/query.service.ts]: I don't think the intention here is to substract the `on` properties.

File: src/http/services/query.service.ts [hunk 3]
```diff
@@ -65,6 +65,17 @@ export class QueryService {
...
+				if (device.fanSpeedControl) {
+					state.currentFanSpeedSetting = device.state.currentFanSeedSetting || 'off';
```
Comment (andrei-tatar) [inline src/http/services/query.service.ts]: The state seems to be missing `currentFanSeedSetting` property. It seems to only have `availableFanSpeeds`.

File: src/models/states/fanspeed.ts [hunk 4]
```diff
@@ -0,0 +1,7 @@
+import { State } from './state';
+
+export type availableFanSpeeds = 'off' | 'low' | 'medium' | 'high';
+
+export interface FanSpeedState extends State {
+     availableFanSpeeds: availableFanSpeeds;
```
Comment (andrei-tatar) [inline src/models/states/fanspeed.ts]: Shouldn't `availableFanSpeeds` be an array? I would also let the node-red plugin to define other values as well.

File: src/http/services/sync.service.ts [hunk 5]
```diff
@@ -93,6 +93,18 @@ export class SyncService {
...
+				case 'lock':
+					sync.type = DeviceTypes.Lock;
+					sync.traits.push(Trait.LockUnlock);
+					break;
+				case "fan":
+					sync.type = DeviceTypes.Fan
+					sync.traits.push(Trait.OnOff);
+					if (device.fanSpeedControl){
+						sync.traits.push(Trait.FanSpeed);
+						sync.attributes = { availableFanSpeeds: uniq(device.availableFanSpeeds.join(',')};
```
Comment (andrei-tatar) [inline src/http/services/sync.service.ts]: from the documentation, availableFanSpeeds is an array of objects that contains name, values, etc.
https://developers.google.com/assistant/smarthome/traits/fanspeed#device-attributes

File: src/models/states/lockunlock.ts [hunk 6]
```diff
@@ -0,0 +1,5 @@
+import { State } from './state';
+
+export interface LockUnlockState extends State {
+    on: boolean;
```
Comment (andrei-tatar) [inline src/models/states/lockunlock.ts]: From the documentation, it seems the states are different:
https://developers.google.com/assistant/smarthome/traits/lockunlock#device-states

File: src/http/services/query.service.ts [hunk 7]
```diff
@@ -65,6 +65,17 @@ export class QueryService {
             case 'garage':
                 state.openPercent = device.state.openPercent;
                 break;
+			case 'lock':
+				state.isLocked = device.state.isLocked;
```
Comment (andrei-tatar) [inline src/http/services/query.service.ts]: `LockUnlockState` is missing these properties.

File: src/http/services/execute.service.ts [hunk 8]
```diff
@@ -47,10 +47,12 @@ export class ExecuteService {
         switch (execution.command) {
           case ExecuteCommandTypes.Brightness:
           case ExecuteCommandTypes.OnOff:
+		  case ExecuteCommandTypes.LockUnlock:
```
Comment (andrei-tatar) [inline src/http/services/execute.service.ts]: `LockUnlock` command should be handled differently. It comes with a `lock` param that should be mapped to `isLocked` state property.
https://developers.google.com/assistant/smarthome/traits/lockunlock#device-commands

**Review by andrei-tatar — CHANGES_REQUESTED** [review]
Needs a few changes. Please try to build the code before submitting a pull request.
Also, please use spaces, not tabs.

**Comment by rgerrans** [comment]
@andrei-tatar So I took a couple steps back and removed all the fan stuff to just focus on the lock now that i spent the time and think I have a better understanding of how the files interconnect and map to the Google pieces. I also took a stab at fixing the execute.service.ts mapping of the Lock  value.

Can you take a look when you have a chance to see if I'm on the right path?

I tried to build these but got a bunch of errors including reference to '@andrei-tatar/ts-ioc'. so guessing I can't setup a build environment?

Thanks for the patience as I learn this stuff.

**Comment by andrei-tatar** [comment]
@rgerrans fixed the `execute.service` for the lock and merged it to master with this PR: https://github.com/andrei-tatar/nora-service/pull/9

**Pull request closed without merging**